core/shared/platform/windows/win_file.c

```diff
@@ -1,115 +1,779 @@
+        return __WASI_FILETYPE_DIRECTORY;
+    }
+    return __WASI_FILETYPE_REGULAR_FILE;
+}
+
+static __wasi_filetype_t
+get_socket_filetype(SOCKET socket)
+{
+    char socket_type = 0;
+    int size = sizeof(socket_type);
+
+    if (getsockopt(socket, SOL_SOCKET, SO_TYPE, &socket_type, &size) == 0) {
+        switch (socket_type) {
+            case SOCK_STREAM:
+                return __WASI_FILETYPE_SOCKET_STREAM;
+            case SOCK_DGRAM:
+                return __WASI_FILETYPE_SOCKET_DGRAM;
+        }
+    }
+    return __WASI_FILETYPE_UNKNOWN;
+}
+
+static __wasi_errno_t
+convert_windows_filetype(os_handle handle, DWORD filetype,
+                         __wasi_filetype_t *out_filetype)
+{
+    __wasi_errno_t error = __WASI_ESUCCESS;
+
+    switch (filetype) {
+        case FILE_TYPE_DISK:
+            FILE_ATTRIBUTE_TAG_INFO file_info;
+
+            bool success = GetFileInformationByHandleEx(
+                handle->raw.handle, FileAttributeTagInfo, &file_info,
+                sizeof(file_info));
+
+            if (!success
+                || file_info.FileAttributes == INVALID_FILE_ATTRIBUTES) {
+                error = convert_windows_error_code(GetLastError());
+                break;
+            }
+
+            *out_filetype = get_disk_filetype(file_info.FileAttributes);
+            break;
+        case FILE_TYPE_CHAR:
+            *out_filetype = __WASI_FILETYPE_CHARACTER_DEVICE;
+            break;
+        case FILE_TYPE_PIPE:
+            if (handle->type == windows_handle_type_socket) {
+                *out_filetype = get_socket_filetype(handle->raw.socket);
+            }
+            else {
+                *out_filetype = __WASI_FILETYPE_BLOCK_DEVICE;
+            }
+            break;
+        case FILE_TYPE_REMOTE:
+        case FILE_TYPE_UNKNOWN:
+        default:
+            *out_filetype = __WASI_FILETYPE_UNKNOWN;
+    }
+
+    return error;
+}
+
+static DWORD
+get_windows_filetype(os_handle handle)
+{
+    switch (handle->type) {
+        case windows_handle_type_normal:
+            return FILE_TYPE_DISK;
+        case windows_handle_type_socket:
+            return FILE_TYPE_PIPE;
+        case windows_handle_type_stdio:
+            return GetFileType(handle->raw.handle);
+        default:
+            return FILE_TYPE_UNKNOWN;
+    }
+}
+
+// Converts the input string to a wchar string.
+static __wasi_errno_t
+convert_to_wchar(const char *str, wchar_t *buf, size_t buf_size)
+{
+    int converted_chars =
+        MultiByteToWideChar(CP_UTF8, 0, str, -1, buf, (int)buf_size);
+
+    if (converted_chars == 0)
+        return convert_windows_error_code(GetLastError());
+
+    return __WASI_ESUCCESS;
+}
+
+// Get the filepath for a handle. The size of the buffer should in terms of
+// wchar.
+static __wasi_errno_t
+get_handle_filepath(HANDLE handle, wchar_t *buf, DWORD buf_size)
+{
+    DWORD bufsize_in_chars = buf_size * (sizeof(wchar_t) / sizeof(char));
+    DWORD size = GetFinalPathNameByHandleW(
```

Comment:
Given that WASI uses non-wide characters, should we just use *A versions of those functions, e.g. `GetFinalPathNameByHandleA`? What benefits are we getting by operating on wide characters?

Reply:
The newer windows APIs only accept wide character strings e.g. CreateFile2. For the sake of consistency, I've just used the wide character equivalents for each Win API function.

From https://learn.microsoft.com/en-us/windows/win32/learnwin32/working-with-strings, `The ANSI versions are also less efficient, because the operating system must convert the ANSI strings to Unicode at run time.` so there shouldn't be any performance disadvantage to using *W functions since the *A functions will convert to wide char internally anyway.